be/src/formats/parquet/level_builder.cpp

```diff
@@ -0,0 +1,708 @@
+    // <list-repetition> group <name> (LIST) {
+    //     repeated group list {
+    //             <element-repetition> <element-type> element;
+    //     }
+    // }
+
+    DCHECK(type_desc.type == TYPE_ARRAY);
+    auto outer_node = std::static_pointer_cast<::parquet::schema::GroupNode>(node);
+    auto mid_node = std::static_pointer_cast<::parquet::schema::GroupNode>(outer_node->field(0));
+    auto inner_node = mid_node->field(0);
+
+    const auto null_col = get_raw_null_column(col);
+    const auto array_col = down_cast<ArrayColumn*>(ColumnHelper::get_data_column(col.get()));
+    const auto& elements = array_col->elements_column();
+    const auto& offsets = array_col->offsets_column()->get_data();
+
+    size_t num_levels_upper_bound = ctx._num_levels + elements->size();
+    auto def_levels = std::make_shared<std::vector<int16_t>>(num_levels_upper_bound,
+                                                             ctx._max_def_level + node->is_optional() + 1);
+    auto rep_levels = std::make_shared<std::vector<int16_t>>(num_levels_upper_bound, ctx._max_rep_level + 1);
+
+    size_t num_levels = 0; // pointer to def/rep levels
+    int offset = 0;        // pointer to current column
+    for (auto i = 0; i < ctx._num_levels; i++) {
+        auto def_level = ctx._def_levels ? (*ctx._def_levels)[i] : 0;
+        auto rep_level = ctx._rep_levels ? (*ctx._rep_levels)[i] : 0;
+
+        // already null in parent column
+        if (def_level < ctx._max_def_level) {
+            (*def_levels)[num_levels] = def_level;
+            (*rep_levels)[num_levels] = rep_level;
+
+            num_levels++;
+            continue;
+        }
+
+        // null in current array_column
+        if (null_col != nullptr && null_col[offset]) {
+            (*def_levels)[num_levels] = def_level;
+            (*rep_levels)[num_levels] = rep_level;
+
+            num_levels++;
+            offset++;
+            continue;
+        }
+
+        auto array_size = offsets[offset + 1] - offsets[offset];
+        // not null but empty array
+        if (array_size == 0) {
+            (*def_levels)[num_levels] = def_level + node->is_optional();
+            (*rep_levels)[num_levels] = rep_level;
+
+            num_levels++;
+            offset++;
+            continue;
+        }
+
+        // not null and non-empty array
+        (*rep_levels)[num_levels] = rep_level;
+        num_levels += array_size;
+        offset++;
+    }
+
+    DCHECK(col->size() == offset);
+
+    def_levels->resize(num_levels);
+    rep_levels->resize(num_levels);
+    LevelBuilderContext derived_ctx(def_levels->size(), def_levels, ctx._max_def_level + node->is_optional() + 1,
+                                    rep_levels, ctx._max_rep_level + 1);
+
+    _write_column_chunk(derived_ctx, type_desc.children[0], inner_node, elements, write_leaf_callback);
+}
+
+void LevelBuilder::_write_array_column_chunk_branchless(const LevelBuilderContext& ctx, const TypeDescriptor& type_desc,
```

Comment:
what's this? remote it ?

Reply:
want to keep it for reference in case of poor readability